warehouse/utils/security_policy.py

```diff
@@ -0,0 +1,120 @@
+from warehouse.packaging.models import Project
+
+
+class AuthenticationMethod(enum.Enum):
+    BASIC_AUTH = "basic-auth"
+    SESSION = "session"
+    MACAROON = "macaroon"
+
+
+def principals_for_authenticated_user(userid, request):
+    """Apply the necessary principals to the authenticated user"""
+    login_service = request.find_service(IUserService, context=None)
+    user = login_service.get_user(userid)
+
+    if user is None:
+        return
+
+    if request.session.password_outdated(login_service.get_password_timestamp(userid)):
+        request.session.invalidate()
+        request.session.flash("Session invalidated by password change", queue="error")
+        return
+
+    principals = []
+
+    if user.is_superuser:
+        principals.append("group:admins")
+    if user.is_moderator or user.is_superuser:
+        principals.append("group:moderators")
+    if user.is_psf_staff or user.is_superuser:
+        principals.append("group:psf_staff")
+
+    # user must have base admin access if any admin permission
+    if principals:
+        principals.append("group:with_admin_dashboard_access")
+
+    return principals
+
+
+@implementer(ISecurityPolicy)
+class MultiSecurityPolicy:
+    """
+    A wrapper for multiple Pyramid 2.0-style "security policies", which replace
+    Pyramid 1.0's separate AuthN and AuthZ APIs.
+
+    Security policies are checked in the order provided during initialization,
+    with the following semantics:
+
+    * `identity`: Selected from the first policy to return non-`None`
+    * `authenticated_userid`: Selected from the request identity, if present
+    * `forget`: Combined from all policies
+    * `remember`: Combined from all policies
+    * `permits`: Uses the AuthZ policy passed during initialization
+
+    These semantics mostly mirror those of `pyramid-multiauth`.
+    """
+
+    def __init__(self, policies, authz):
+        self._policies = policies
+        self._authz = authz
+
+    def identity(self, request):
+        for policy in self._policies:
+            if ident := policy.identity(request):
+                return ident
+
+        return None
+
+    def authenticated_userid(self, request):
+        if request.identity:
+            return str(request.identity.id)
+        return None
+
+    def forget(self, request, **kw):
+        headers = []
+        for policy in self._policies:
+            headers.extend(policy.forget(request, **kw))
+        return headers
+
+    def remember(self, request, userid, **kw):
+        headers = []
+        for policy in self._policies:
+            headers.extend(policy.remember(request, userid, **kw))
+        return headers
+
+    def permits(self, request, context, permission):
+        identity = request.identity
+        principals = []
+        if identity is not None:
+            # Our identity is always an ORM model, so add some generic principals first.
+            principals.extend([Authenticated, str(identity.id)])
```

Comment:
> Writing the id in the principals seems slightly dangerous to me because there could be a confusion if a user and a project had the same id. I know it's very very unlikely due to both ids being UUIDs, but I don't think it's a good idea to receive an ID and to have to guess what model it belongs to.

I might be misunderstanding what you mean, but we should never actually have to guess the model -- `request.identity` will always be the correct ORM model, so the `id` here is solely for permission matching.

(That being said, your suggestion re: `f"user:{user.id}"` sounds reasonable to me, if we do decide that raw IDs are ambiguous.)

Reply:
I think what @ewjoachim is that it's possible (but unlikely) for an `id` to be shared by both a `User` and a `Project`, which I think is correct.

I'm not seeing why we're doing this though, were we doing it before and I'm just missing it? Or is this something taken from the new 2.0-style examples? (I'm not sure it's necessary).

Reply:
Yeah, it's another thing I cargo-culted from the upgrade guide. I can probably remove it entirely with nothing breaking, will experiment now.

Reply:
Hmm. I tried to remove this and things *appear* to work, but package upload with HTTP basic auth always 403s without it. Any thoughts?

Reply:
https://github.com/pypa/warehouse/blob/main/warehouse/packaging/models.py#L217-L236

Reply:
Thanks @dstufft! That's an easy fix.

Reply:
Yep, that's it. So I think @ewjoachim's idea to prefix this with `user:` makes sense.

Reply:
@dstufft do you happen to know how the original code was placing the user ID in the principals list? I'm having trouble finding it.

Reply:
https://github.com/pypa/warehouse/blob/main/warehouse/accounts/__init__.py#L66-L92

Reply:
I don't remember why I used strings there instead of ORM objects, there might have been a reason, but it might also work just fine to use objects. I think probably the examples all used strings, and I didn't feel like digging into it to figure out if not-strings would introduce some horrible security bug.

Reply:
I basically preserved that function as `_principals_for_authenticated_user`, but I don't actually see the user's ID getting added to the principal set there. Unless I'm missing something?

Reply:
Oh right. Sorry, the ``request.effective_principals`` in Pyramid pre 2.0 (no idea what happens after that) gets populated by the authentication policy's ``effective_principals`` method. Both the Session and the Basic auth authN policy that ships (shipped?) with Pyramid used the ``CallbackAuthenticationPolicy``, which looks like:

https://github.com/Pylons/pyramid/blob/1.9-branch/pyramid/authentication.py#L107-L184

Which puts the special "Everyone" ID, the string of the user id, and then anything that comes from the "groups" callback, which is our ``_authenticate`` method.

It looks like the Macaroon policy also uses the ``CallbackAuthenticationPolicy``, so it will also do that.

Reply:
Mystery solved! 😄 

Yeah, we've gotten rid of the callbacks entirely (since they're not part of the 2.0 APIs), so now everything is concentrated into that one `permits` method that dispatches into the underlying stack of AuthZ policies.

Reply:
Assuming something like this still exists in Pyramid 2.0, we could just ignore the fact there is a string-ified userid in the principals, and make the ``_authenticate`` method also return a ``f"{model}:{id}"`` prinicpal, and actually use that in our ACLs OR we can override some methods to make our own ``CallbackAuthenticationPolicy`` that has the behavior we want.

Reply:
Ah cool, that's easier then.